File: container_depot/operations/report/container_siap_keluar/container_siap_keluar.js
// Container Siap Keluar — reminder queue between a clean EIR-Out and the tank actually
// leaving. The "ACC Keluar" cell is a button rather than a value: the Desk user confirms the
// exit from the same list they read it in, and it calls the very endpoint the PWA calls
// (container_depot.ess.gate.gate_out), so both surfaces enforce the same guards.

frappe.query_reports["Container Siap Keluar"] = {
	filters: [
		{
			fieldname: "search",
			label: __("Cari"),
			fieldtype: "Data",
			description: __("No. container / bon muat / truk / sopir"),
		},
	],

	formatter(value, row, column, data, default_formatter) {
		if (column.fieldname === "acc") {
			if (!data || !data.container) return "";
			return `<button class="btn btn-xs btn-primary csk-acc" data-container="${frappe.utils.escape_html(
				data.container
			)}" data-label="${frappe.utils.escape_html(data.container_no || data.container)}">${__(
				"ACC Keluar"
			)}</button>`;
		}
		// A tank standing ready for days is the thing this report exists to surface.
		if (column.fieldname === "waiting" && data && /hari/.test(value || "")) {
			return `<span class="text-danger font-weight-bold">${frappe.utils.escape_html(value)}</span>`;
		}
		return default_formatter(value, row, column, data);
	},

	onload(report) {
		// Delegated so it survives every datatable re-render; namespaced so a report
		// refresh never stacks duplicate handlers (which would fire gate-out twice).
		$(report.page.wrapper)
			.off("click.cskAcc")
			.on("click.cskAcc", ".csk-acc", function () {
				const container = $(this).data("container");
				const label = $(this).data("label") || container;
				frappe.confirm(
					__("Konfirmasi {0} keluar depo (muat selesai)? Tindakan ini tidak bisa dibatalkan.", [
						`<b>${frappe.utils.escape_html(label)}</b>`,
					]),
					() => {
						frappe.call({
							method: "container_depot.ess.gate.gate_out",
							args: { container },
							freeze: true,
							freeze_message: __("Memproses gate-out …"),
							callback: (r) => {
								if (!r || r.exc) return;
								const done = r.message && r.message.order_completed;
								frappe.show_alert({
									message: done
										? __("{0} keluar — bon {1} selesai", [label, r.message.order_muat])
										: __("{0} keluar depo — gate-out selesai", [label]),
									indicator: "green",
								});
								report.refresh();
							},
						});
					}
				);
			});
	},
};

File: container_depot/container_depot/report/container_siap_keluar/container_siap_keluar.py
"""Container Siap Keluar — the Desk twin of the PWA "Siap Keluar" worklist.

Every tank whose EIR-Out is submitted clean but that is still standing in the depot: the
reminder queue between the surveyor finishing and the truck actually leaving. The rows come
straight from :func:`container_depot.container_depot.gate.list_ready_to_load` so the Desk and the
PWA can never disagree, and the "ACC Keluar" button in the last column calls the same
``gate_out`` endpoint the PWA does (see ``container_siap_keluar.js``).
"""

from frappe.utils import get_datetime, now_datetime

from container_depot.container_depot.gate import list_ready_to_load


def execute(filters=None):
	return get_columns(), get_data(filters)


def get_columns():
	return [
		{"fieldname": "container", "label": "Container", "fieldtype": "Link", "options": "Container", "width": 130},
		{"fieldname": "order_muat", "label": "Bon Muat", "fieldtype": "Link", "options": "Order Muat", "width": 140},
		{"fieldname": "waiting", "label": "Menunggu", "fieldtype": "Data", "width": 100},
		{"fieldname": "ready_since", "label": "Siap Sejak", "fieldtype": "Datetime", "width": 160},
		{"fieldname": "truck_no", "label": "Truk", "fieldtype": "Data", "width": 110},
		{"fieldname": "driver", "label": "Sopir", "fieldtype": "Data", "width": 130},
		{"fieldname": "destination", "label": "Tujuan", "fieldtype": "Data", "width": 140},
		{"fieldname": "principal", "label": "Principal", "fieldtype": "Data", "width": 130},
		{"fieldname": "depot", "label": "Depo", "fieldtype": "Data", "width": 110},
		{"fieldname": "inspection", "label": "EIR-Out", "fieldtype": "Link", "options": "Inspection", "width": 140},
		{"fieldname": "acc", "label": "Aksi", "fieldtype": "Data", "width": 120},
	]


def get_data(filters=None):
	filters = filters or {}
	rows = list_ready_to_load(
		search=filters.get("search"), start=0, page_length=500
	)["items"]
	now = now_datetime()
	for row in rows:
		row["waiting"] = _humanize_wait(row.get("ready_since"), now)
		row["acc"] = ""  # rendered as a button by the report's formatter
	return rows


def _humanize_wait(ready_since, now):
	"""How long the tank has been standing ready — the whole point of the reminder."""
	if not ready_since:
		return ""
	hours = (now - get_datetime(ready_since)).total_seconds() / 3600
	if hours < 1:
		return f"{int(hours * 60)} menit"
	if hours < 24:
		return f"{int(hours)} jam"
	return f"{int(hours // 24)} hari"
